bloom-offchain/src/execution_engine/mod.rs

```diff
@@ -372,19 +345,16 @@ where
         CO: Stable<StableId = SID> + Clone + Display,
         P: Stable<StableId = SID> + Clone,
         IX: StateIndex<EvolvingEntity<CO, P, V, B>>,
-        CH: KvStore<SID, EvolvingEntity<CO, P, V, B>>,
         TLB: ExternalLBEvents<CO, P> + Maker<MC>,
     {
         for ver in versions {
-            if let Some(stable_id) = self.index.invalidate_version(ver) {
+            if let Some(state) = self.index.get_state(ver) {
+                let stable_id = state.stable_id();
+                let state_before_invalidation = resolve_state(stable_id, &self.index);
                 trace!("Invalidating snapshot {} of {}", ver, stable_id);
-                let maybe_transition = match resolve_source_state(stable_id, &self.index) {
-                    None => self
-                        .cache
-                        .remove(stable_id)
-                        .map(|Bundled(elim_state, _)| Ior::Left(elim_state)),
-                    Some(latest_state) => self.cache(latest_state),
-                };
+                self.index.invalidate_version(ver);
+                let state_after_invalidation = resolve_state(stable_id, &self.index);
+                let maybe_transition = to_transition(state_before_invalidation, state_after_invalidation);
```

Comment:
If we are invalidating versions that should be presented in `index`, how it's possible that `to_transition()` will return `(None, None)`?